### PR TITLE
eagerly validate arguments; only accept iterables

### PR DESCRIPTION
**Pull request by michaelficarra**
Fixes #9.

### CODE REVIEW

**Comment by bakkot** [comment]
I think you probably don't want to open iterables that you never get to, which makes the logic a bit more complicated: you need to store either `Symbol.iterator` or `next`, and keep track of which is which so you know whether to close.

**Comment by michaelficarra** [comment]
@bakkot You wanna push up what you're talking about?

**Comment by michaelficarra** [comment]
@bakkot Please take a look and let me know if this is what you were thinking.

File: spec.emu [hunk 1]
```diff
@@ -12,11 +12,24 @@ copyright: false
...
-      1. Repeat, while _items_ is not empty,
-        1. Let _iter_ be the first element of _items_.
-        1. Remove the first element from _items_.
-        1. Let _iteratorRecord_ be ? GetIteratorFlattenable(_iter_, ~reject-strings~).
+    1. Let _iteratorRecords_ be a new empty List.
+    1. For each element _item_ of _items_, do
+      1. If _item_ is not an Object, throw a *TypeError* exception.
+      1. Let _method_ be ? GetMethod(_item_, %Symbol.iterator%).
+      1. If _method_ is *undefined*, then
+        1. Append ? GetIteratorDirect(_item_) to _iteratorRecords_.
```
Comment (bakkot) [inline spec.emu]:
```suggestion
        1. Let _v_ be ? GetIteratorDirect(_item_) to _iteratorRecords_.
        1. If IsCallable(_v_) is *false*, throw a *TypeError* exception.
        1. Append _v_ to _iteratorRecords_.
```

**Comment by bakkot** [comment]
Other than my above comment, and the handling of early exits as described below, yes, that's what I meant.

Handling of early exits is tricky: most iterators are themselves iterable, so you can't assume that just because it has a `Symbol.iterator` property it can't be an iterator, and we do need to ensure we close all the iterators we're given. Practically speaking I think this means you need to check for `return` on every not-yet-consumed argument, regardless of which kind of thing it is.

And you also need to do this for exceptions which arise when first walking through the list of arguments as well: `Iterator.concat(openIterableIterator1, null, openIterableIterator2)` should close both `openIterableIterator1` and `openIterableIterator2`.

But those fixes can be done in a followup PR.

**Comment by michaelficarra** [comment]
I've updated the PR to just accept iterables as arguments. Open iterators that are not consumed will not be closed.

**Review by bakkot — APPROVED** [review]
WFM

**Pull request merged**